### PR TITLE
Enable ETCD with TLS

### PR DESCRIPTION
**Pull request by nolancon**
## PR Explainer/Overview

This PR, along with https://github.com/storageos/etcd-cluster-operator/commit/8829e7a39889a5edaf096b6de4354bb71431bb8a and https://github.com/storageos/etcd-cluster-operator/commit/85974f4555df05649c2bf7df34a57202e00f7bc6 allows the plugin to install ETCD and storageos with TLS configured.

### Installing **both** ETCD and StorageOS with the `--include-etcd` flag:
In this scenario, by setting the flag `--etcd-tls-enabled`, the installer does the following:

1. Sets `tls.enabled: true` in the EtcdCluster spec
2. Sets `tls.storageOSClusterNamespace: <storageos-cluster-namespace>` in the EtcdCluster spec
3. Sets `tls.storageOSEtcdSecretName: <etcd-secret-name>` in the EtcdCluster spec. This can optionally be changed by setting the (new) `--etcd-secret-name` install flag (default = `storageos-etcd-secret`).

When the above fields are set in the EtcdCluster by the installer prior to ETCD installation, the etcd-operator will create an additional k8s secret (etcd-secret-name) in the storageos cluster namespace. This secret will be used by storageos to access ETCD over TLS.

4. Sets `tlsEtcdSecretRefName: <etcd-secret-name>`  in the StorageOSCluster spec. This can optionally be changed by setting the (new) `--etcd-secret-name` install flag (default = `storageos-etcd-secret`).
5.  Sets `tlsEtcdSecreRefNamespace: <storageos-cluster-namespace>` in the StorageOSCluster spec

Once installation is complete, StorageOS should be working as expected with ETCD over TLS

#### Full command to test:
`kubectl storageos install --include-etcd --etcd-tls-enabled`

#### Full command to test (with optionally named secret):
`kubectl storageos install --include-etcd --etcd-tls-enabled --etcd-secret-name=test-secret`

### Installing  StorageOS only
1. Sets `tlsEtcdSecretRefName: <etcd-secret-name>` in the StorageOSCluster spec. This can optionally be changed by setting the (new) `--etcd-secret-name` install flag (default = `storageos-etcd-secret`).
2.  Sets `tlsEtcdSecreRefNamespace: <storageos-cluster-namespace>` in the StorageOSCluster spec.

The installer then searches for secret (etcd-secret-name) in the storageos cluster namespace.
If the secret is not found, the user is presented with an error message detailing how to create a valid secret in order to enable TLS (see [here](https://github.com/nolancon/kubectl-storageos/blob/a53229fae5e1335cfd8468ed27630fca69abcfd4/pkg/installer/endpoints.go#L20))

If the secret is found (i.e if it has been created manually by the user **or** it exists from a previous etcd installation by the plugin) the installer creates a static etcd-shell-pod (existing behaviour, but this time...) with the secret mounted. Now the plugin will attempt to validate the user-defined etcd endpoints over HTTPS using the static pod and secret. If successful, the installation continues as before.

Etcd endpoint validation can also be bypassed completely by setting the (new) `--skip-etcd-endpoint-validation` flag (false by default).

#### Full command to test:
`kubectl storageos install --etcd-tls-enabled`

#### Full command to test (with optionally named secret):
`kubectl storageos install --etcd-tls-enabled --etcd-secret-name=test-secret`

#### Full command to test (with no endpoint validation - not recommended):
`kubectl storageos install --etcd-tls-enabled --skip-etcd-endpoint-validation`

## TODOs:

- [x] Fix e2e tests

### CODE REVIEW

File: pkg/installer/endpoints.go [hunk 1]
```diff
@@ -4,34 +4,83 @@ import (
...
 	"github.com/storageos/kubectl-storageos/pkg/logger"
+
 	pluginutils "github.com/storageos/kubectl-storageos/pkg/utils"
-	"k8s.io/client-go/rest"
 )
 
 const (
-	etcdShellUrl = "https://raw.githubusercontent.com/nolancon/placeholder/main/storageos-etcd-shell.yaml"
-	httpsPrefix  = "https://"
+	nSForSecretNotFoundError = `
```
Comment (mhmxs) [inline pkg/installer/endpoints.go:15]: Idiomatically `errNSForSecretNotFound` should be a better name

File: pkg/installer/endpoints.go [hunk 2]
```diff
@@ -4,34 +4,83 @@ import (
...
+	Namespace %s not found for %s while attempting to validate ETCD endpoints.
+
+	To skip ETCD endpoints validation during installation, set the --%s flag.
+`
+	secretNotFoundError = `
```
Comment (mhmxs) [inline pkg/installer/endpoints.go:20]: Same here.

File: pkg/installer/endpoints.go [hunk 3]
```diff
@@ -4,34 +4,83 @@ import (
...
-// - adds validated endpoints patch to kustomization file for storageos-cluster.yaml
-func (in *Installer) handleEndpointsInput(etcdEndpoints string) error {
-	fmt.Println("Warning: TLS endpoints are not supported")
-	if strings.HasPrefix(etcdEndpoints, httpsPrefix) {
-		return fmt.Errorf("TLS endpoint discovered")
+// handleEndpointsInput adds validated (or not validated) endpoints patch to kustomization file
+// for storageos-cluster.yaml
+func (in *Installer) handleEndpointsInput(configInstall apiv1.Install) error {
+	if !configInstall.SkipEtcdEndpointsValidation {
+		err := in.validateEtcd(configInstall)
```
Comment (mhmxs) [inline pkg/installer/endpoints.go:45]: Single error could be handled inside the `if`

File: pkg/installer/endpoints.go [hunk 4]
```diff
@@ -4,34 +4,83 @@ import (
...
+		Value: configInstall.EtcdEndpoints,
+	}
+
+	fsClusterName, err := in.getFieldInFsMultiDocByKind(filepath.Join(stosDir, clusterDir, stosClusterFile), stosClusterKind, "metadata", "name")
+	if err != nil {
+		return err
 	}
 
-	etcdShell, err := pullManifest(etcdShellUrl)
+	err = in.addPatchesToFSKustomize(filepath.Join(stosDir, clusterDir, kustomizationFile), stosClusterKind, fsClusterName, []pluginutils.KustomizePatch{endpointPatch})
```
Comment (mhmxs) [inline pkg/installer/endpoints.go:61]: Single error could be handled inside the if

File: pkg/installer/endpoints.go [hunk 5]
```diff
@@ -4,34 +4,83 @@ import (
...
 	if err != nil {
```
Comment (mhmxs) [inline pkg/installer/endpoints.go:62]: If err not nil, return err, but if nil return nil. Why not simply return err?

File: pkg/installer/endpoints.go [hunk 6]
```diff
@@ -45,36 +94,59 @@ func (in *Installer) handleEndpointsInput(etcdEndpoints string) error {
 		return nil
 	}()
 
-	err = validateEndpoints(etcdEndpoints, string(etcdShell))
+	err = in.validateEndpoints(configInstall.EtcdEndpoints, string(etcdShell), configInstall.EtcdTLSEnabled)
```
Comment (mhmxs) [inline pkg/installer/endpoints.go:97]: Single error could be handled inside the if

File: pkg/installer/endpoints.go [hunk 7]
```diff
@@ -45,36 +94,59 @@ func (in *Installer) handleEndpointsInput(etcdEndpoints string) error {
...
 	if err != nil {
 		return err
 	}
+	return nil
```
Comment (mhmxs) [inline pkg/installer/endpoints.go:101]: Why not simply return err?

File: pkg/installer/endpoints.go [hunk 8]
```diff
@@ -45,36 +94,59 @@ func (in *Installer) handleEndpointsInput(etcdEndpoints string) error {
...
-	endpointPatch := pluginutils.KustomizePatch{
-		Op:    "replace",
-		Path:  "/spec/kvBackend/address",
-		Value: etcdEndpoints,
+// tlsValidationPrep:
+// - searches for the etcd-secret
+// - applies app=storageos label to secret
+// - returns the tls equipped etcd-shell pod with storageos cluster namespace and secret name
+func (in *Installer) tlsValidationPrep(configInstall apiv1.Install) (string, error) {
+	err := pluginutils.NamespaceExists(in.clientConfig, configInstall.StorageOSClusterNamespace)
```
Comment (mhmxs) [inline pkg/installer/endpoints.go:109]: Single error could be handled inside the if

File: pkg/installer/endpoints.go [hunk 9]
```diff
@@ -45,36 +94,59 @@ func (in *Installer) handleEndpointsInput(etcdEndpoints string) error {
...
+	// apply app=storageos label to secret, this way it will be backed up locally during uninstall
+	secretLabels := etcdSecret.GetLabels()
+	secretLabels["app"] = "storageos"
+	etcdSecret.SetLabels(secretLabels)
+	etcdSecretManifest, err := secretToManifest(etcdSecret)
 	if err != nil {
-		return err
+		return "", err
+	}
+	err = in.kubectlClient.Apply(context.TODO(), configInstall.StorageOSClusterNamespace, string(etcdSecretManifest), true)
```
Comment (mhmxs) [inline pkg/installer/endpoints.go:126]: Single error could be handled inside the if

File: pkg/installer/endpoints.go [hunk 10]
```diff
@@ -137,3 +205,47 @@ func etcdctlHealthCheck(config *rest.Config, etcdShellPodName, etcdShellPodNS, e
 
 	return nil
 }
+
+func etcdShellPodTLS() string {
```
Comment (mhmxs) [inline pkg/installer/endpoints.go:209]: What is the reason of a function instead of constant?

Comment (nolancon) [reply pkg/installer/endpoints.go]: no good reason 😄 changed to const

**Review by mhmxs — APPROVED** [review]

**Pull request merged**